django/db/models/query.py

```diff
@@ -813,7 +813,12 @@ def bulk_create(
         objs = list(objs)
         self._prepare_for_bulk_create(objs)
         with transaction.atomic(using=self.db, savepoint=False):
-            objs_with_pk, objs_without_pk = partition(lambda o: o.pk is None, objs)
+            objs_with_pk, objs_without_pk = partition(
+                lambda o: (
+                    o.pk is None or (hasattr(o.pk, "is_set") and not o.pk.is_set())
+                ),
+                objs,
+            )
```

Comment:
I find the `lambda` here pretty confusing. I think defining a full helper function is easier to understand:

```suggestion
            def pk_is_unset(obj):
                if obj.pk is None:
                    return True
                try:
                    return not obj.pk.is_set()
                except AttributeError:
                    return False
            
            objs_with_pk, objs_without_pk = partition(pk_is_unset, objs)
```

It might even be better to invert the return value:

```suggestion
            def pk_is_set(obj):
                pk = obj.pk
                try:
                    return pk.is_set()
                except AttributeError:
                    return pk is not None
            
            objs_without_pk, objs_with_pk = partition(pk_is_set, objs)
```